### PR TITLE
Follow up of VM creation redesign

### PR DESCRIPTION
**Pull request by skobyda**
Split from: https://github.com/cockpit-project/cockpit-machines/pull/560

### CODE REVIEW

**Comment by skobyda** [comment]
In previous PR, @martinpitt had few issues with naming. @garrett might wanna join discussion too:

martin wrote:
`It's redundant to include the connection name into the domain name. These are already namespaced, and it will make them even longer.`

Ah that's true. If we don't include them, we will just end up with VM list looking like this. I'm fine removing system/session from the name:
Screenshot from 2022-04-11 11-47-51


martin wrote:
```
"a name will be suggested" is too surprising. This would only appear until the user selects an OS, right? and that is mandatory? We need to avoid a situation where you click "create" with "a name will be suggested".

Still, I feel this is too magic/jumpy, as a helper text. Easier/less surprising behaviour proposal: set the name input to an autogenerated name once the user chooses an OS and the name is empty still, and then just keep the current "is empty" validation.

Also, grammar: If you keep this sentence, it's "If blank, name will be set.." (drop the "a"). But still, just do it, and avoid the indirection.

As this whole auto-naming business is quite hairy, I also recommend splitting this out into its own PR. I feel it's getting unnecessarily difficult to handle all these independent changes in a single PR.
```

**Comment by skobyda** [comment]
> It's redundant to include the connection name into the domain name. These are already namespaced, and it will make them even longer.

Ah that's true. If we don't include them, we will just end up with VM list looking like this. I'm fine removing system/session from the name:
Screenshot from 2022-04-11 11-47-51

**Comment by skobyda** [comment]
> Split from: #560

Screenshot from 2022-04-11 11-51-26
Screenshot from 2022-04-11 11-51-19

**Comment by garrett** [comment]
> Ah that's true. If we don't include them, we will just end up with VM list looking like this. I'm fine removing system/session from the name:
> Screenshot from 2022-04-11 11-47-51

Shouldn't the list also be renamed to "User session", like the toggle in the create dialog? That'd help differentiate between "System" and "Session", which look all too similar here otherwise.

**Comment by skobyda** [comment]
> Shouldn't the list also be renamed to "User session", like the toggle in the create dialog? That'd help differentiate between "System" and "Session", which look all too similar here otherwise.

I thank you, I didn't catch that. Fixed it:

Screenshot from 2022-04-11 18-37-00

**Comment by garrett** [comment]
> "a name will be suggested" is too surprising. This would only appear until the user selects an OS, right?
> ...
> Easier/less surprising behaviour proposal: set the name input to an autogenerated name once the user chooses an OS and the name is empty still, and then just keep the current "is empty" validation.

Let's do this.

The generated names are really best as a fallback. People should be naming their VMs with some purpose, like "SQL server", "DNS", "NGINX", "NextCloud", etc. — which we cannot guess. The generated names as a fallback are good, however, as differentiators that do provide some context about the VM... but they're not as good as a descriptive name. They're definitely better than an error, especially for the cases where someone wants to make a quick VM for testing, so this PR is still a nice feature.

Removing the text and treating this just as a fallback where the name wasn't entered (before selecting configurations) will simplify the UI, would make this PR easier to land, and (as a bonus) make @martinpitt happy too. :wink:

**Comment by skobyda** [comment]
Current state:

### When user opens the dialog, they do not see any message about autogeneration:
Screenshot from 2022-04-20 14-12-19

### Once they enter the Operating system, the message appears:
Screenshot from 2022-04-20 14-12-26

### If afterwards they write some VM name themself, the message disappears:
Screenshot from 2022-04-20 14-12-36

**Comment by garrett** [comment]
I don't like that the form jumps down from above with a setting below the area it jumps.

What happened to this?

>>     Easier/less surprising behaviour proposal: set the name input to an autogenerated name once the user chooses an OS and the name is empty still, and then just keep the current "is empty" validation.
>
> Let's do this.

I guess I didn't _clearly_ specifically specify helper text or not, but I was assuming it didn't. I did say "removing the text", but I didn't make it clear that it was helper text:

> _Removing the text_ and treating this just as a fallback where the name wasn't entered (before selecting configurations) will simplify the UI, would make this PR easier to land, and (as a bonus) make @martinpitt happy too. 

We could, however, do a hybrid approach and use placeholder text, I suppose... as this would be a fallback.

https://www.patternfly.org/v4/components/form/design-guidelines#placeholder-text-0

We could have the placeholder text say "_Unique name_" by default, and I suppose "_Unique name, default: XYZ_" where XYZ is the generated string. (I had to modify the PF recommendation a little to make this work. But it's similar "in spirit" to what the guidelines suggest.)

Doing this would:

- Relegate the autonaming to a fallback
- Not cause a jump in the form and modal height
- Still expose the name that will be chosen if nothing else is specified
- Only show up when nothing else is typed as the name (which happens to be what we want, really, if this is just a fallback)

**Comment by skobyda** [comment]
Ahh, sorry I misunderstood previous comments, I though only "If left blank, name will be autogenerated" message should removed, but "If left blank, name will be set to someName" should stay.
Anyway, I hope now I got it right:
### When user opens the dialog, everything is as before:
Screenshot from 2022-04-25 19-25-37

### If user first specifies VM name, and then selects an OS, no name autogeneration will take place:
Screenshot from 2022-04-25 18-56-23

### If user however doesn't specify VM name, and goes straight to specifying OS:
Screenshot from 2022-04-25 19-14-28

### If user doesn't specify either, and clicks on "Create...", validation still works as before:

Screenshot from 2022-04-25 19-15-01

File: src/components/create-vm-dialog/createVmDialog.jsx [hunk 1]
```diff
@@ -784,9 +803,11 @@ class CreateVmModal extends React.Component {
 
     onValueChanged(key, value) {
         switch (key) {
-        case 'vmName':
-            this.setState({ [key]: value.split(" ").join("_") });
+        case 'vmName': {
+            const stateDelta = { [key]: value.split(" ").join("_") };
+            this.setState(stateDelta);
             break;
+        }
```
Comment (martinpitt) [inline src/components/create-vm-dialog/createVmDialog.jsx:810]: No-op change

File: test/check-machines-create [hunk 2]
```diff
@@ -139,6 +140,49 @@ class TestMachinesCreate(VirtualMachinesCase):
                                                                          create_and_run=False),
                                              {"source-url": "Installation source must not be empty"})
 
+    def testCreateNameGeneration(self):
```
Comment (martinpitt) [inline test/check-machines-create:143]: Please add this to the exclude list in test/browser/run-test.sh , as it tests internal logic only.

**Review by martinpitt — CHANGES_REQUESTED** [review]
Nice, thank you! I like that workflow much more than the "surprise! :tada:" approach. One or two tiny things to clean up, then it's good for me. @garrett, are you d'accord with this workflow, too? From what I can see, it matches your specification exactly. Cheers!

**Review by garrett — CHANGES_REQUESTED** [review]
> Ahh, sorry I misunderstood previous comments

That's fine. The interpretation of what I originally wrote wasn't clear and, reading it again later, I could see it being understood either way.

So, after your recent changes: It looks mostly great! 

~However, I did find an interaction bug:~

~Select download, then select an OS, then switch download to something else. The suggested name does not update. The suggested name should probably be recomputed (if the name is blank) after every form change (possibly without an autogenerated name, if the values don't support it), not just the applicable ones (which is what it seems to currently do).~

~Screenshot after I select download CentOS 7, then switched to cloud base image:~

~Screenshot 2022-04-26 at 14-53-30 Virtual Machines - garrett@Bolt~

Edit: Oh! I still have CentOS 7 selected. Whoops! This is fine.

**Review by garrett — DISMISSED** [review]
:+1: Looks great!

**Comment by martinpitt** [comment]
The [two](https://logs.cockpit-project.org/logs/pull-665-20220426-120759-34fcaffa-rhel-9-0/log.html) [RHEL 9.x](https://logs.cockpit-project.org/logs/pull-665-20220426-120759-34fcaffa-rhel-9-1/log.htmll failures look a bit curious, like there's some VM corruption going on? I've never seen that before, this is worth a closer look. Can you please rebase, to run the packit tests, and retry the ones on our CI?

**Comment by martinpitt** [comment]
Otherwise LGTM, will approve once green. Thanks!

**Comment by martinpitt** [comment]
Rebased (without other changes).

**Comment by skobyda** [comment]
> Meh, there is [something about that test](https://logs.cockpit-project.org/logs/pull-665-20220426-155628-cb51d3a7-rhel-9-0/log.html#72) which makes it very prone to corrupting the VM.

That's very odd. It seems that VM fails even to get prepared. But there is nothing unusual with this test which would fail it

**Comment by martinpitt** [comment]
@skobyda : How do you mean? It was running before, the test killed it. Hence the timeout and "Restarting global machine 6".

**Comment by skobyda** [comment]
Ah I was looking at different test run. @martinpitt after I retriggerred test, the failure is different: https://logs.cockpit-project.org/logs/pull-665-20220427-090744-7c02b2de-rhel-9-0/log.html
It says it failed at bots/machine/machine_core/ssh_connection.py, but isn't that a VM preparation which runs before the test?

**Comment by martinpitt** [comment]
No, VM prep happens in the "initialization" section. If you grep the [raw log](https://logs.cockpit-project.org/logs/pull-665-20220427-090744-7c02b2de-rhel-9-0/log) for "ND@3" you see all tests which run in that persistent VM. `TestMachinesCreate.testCreateNameGeneration` killed it, presumably due to some kernel crash or an out of memory condition. Possibly the created VMs are allocating too much memory? This needs reproducing locally and checking with `virsh console`.

**Comment by skobyda** [comment]
Tests will fail until https://github.com/cockpit-project/cockpit-machines/pull/696 gets merged

**Comment by jelly** [comment]
> Blocked, tests will fail until #696 gets merged

Reviewed and merged so this can be re-tested.

**Review by martinpitt — APPROVED** [review]
Thanks! Enough nitpicking, let's land this.

**Comment by garrett** [comment]
If we're going to mention this in the release notes, it needs a very short summary of what the PR actually changes as the title and a bit more detailed summary as the text at the top, ideally with a screenshot. Thanks!

**Pull request merged**